### PR TITLE
Mutiny API

### PR DESCRIPTION
**Pull request by gavinking**
This work introduced a new parallel API for Mutiny, `Mutiny.Session`, `Mutiny.Query`, etc, and cleans up some mess associated with the `SessionFactory` and `SessionFactoryBuilder`.

See #94.

### CODE REVIEW

**Comment by gavinking** [comment]
@FroMage take a look please and let me know if this is what you need for Panache.

**Comment by gavinking** [comment]
Alright, so I'm done with my rebase/squash and some extra cleanups, so I guess this is ready to merge.

**Comment by DavideD** [comment]
This is awesome!

**Comment by gavinking** [comment]
Merged!

File: hibernate-reactive-core/src/test/java/org/hibernate/reactive/BaseMutinyTest.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+	@Rule
+	public Timeout rule = Timeout.seconds( 3600 );
+
+	private Mutiny.Session session;
+	private ReactiveConnection connection;
+	private org.hibernate.SessionFactory sessionFactory;
+	private ReactiveConnectionPool poolProvider;
+
+	protected static void test(TestContext context, CompletionStage<?> cs) {
```
Comment (FroMage) [inline hibernate-reactive-core/src/test/java/org/hibernate/reactive/BaseMutinyTest.java:35]: I would expect a mutiny test to use `Uni` here, but :shrug: :)

File: hibernate-reactive-core/src/test/java/org/hibernate/reactive/MutinySessionTest.java [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+	}
+//
+//	private Uni<Integer> cleanDB() {
+//		return connection().flatMap( conn -> Uni.createFrom().completionStage( conn.update( "DELETE FROM Pig" ) ) );
+//	}
+
+	private Uni<String> selectNameFromId(Integer id) {
+		return connection().flatMap(
+				connection -> Uni.createFrom().completionStage(
+						connection.select(
```
Comment (FroMage) [inline hibernate-reactive-core/src/test/java/org/hibernate/reactive/MutinySessionTest.java:35]: Is it normal to have to use `CompletionStage` for `select`?

File: hibernate-reactive-core/src/test/java/org/hibernate/reactive/BaseMutinyTest.java (hunk 1 above)
Comment (gavinking) [reply hibernate-reactive-core/src/test/java/org/hibernate/reactive/BaseMutinyTest.java:35]: Yes yes, at this point I had not finished work on the test harness but it is finished now, so if you would be looking at the current code, and not a 4-day-old pull request, you wouldn't need to see things like this :-P

File: hibernate-reactive-core/src/test/java/org/hibernate/reactive/MutinySessionTest.java (hunk 2 above)
Comment (gavinking) [reply hibernate-reactive-core/src/test/java/org/hibernate/reactive/MutinySessionTest.java:35]: There is no Mutiny API for `ReactiveConnection`. It's an internal API, not supposed to be visible to the program. (You have to be an expert in Hibernate `Service`s to even get a reference to it.)

**Comment by gavinking** [comment]
@FroMage this is where you should be looking:

<https://github.com/hibernate/hibernate-reactive/blob/master/example/src/main/java/org/hibernate/example/reactive/MutinyMain.java>

File: hibernate-reactive-core/src/test/java/org/hibernate/reactive/BaseMutinyTest.java (hunk 1 above)
Comment (FroMage) [reply hibernate-reactive-core/src/test/java/org/hibernate/reactive/BaseMutinyTest.java:35]: Well, you raise the PR while I'm on PTO and I'm kind enough to review it even when you didn't wait for me to merge it, so don't complain too much ;)

File: hibernate-reactive-core/src/test/java/org/hibernate/reactive/MutinySessionTest.java (hunk 2 above)
Comment (FroMage) [reply hibernate-reactive-core/src/test/java/org/hibernate/reactive/MutinySessionTest.java:35]: OK, fair.

**Pull request merged**